Add plugin version to GRPC interface

swenson opened the pull request:
Added a version interface in the `sdk/logical` so that it can be shared between all plugin types, and then wired it up to `RunningVersion` in the mounts, auth list, and database systems.

I've tested that this works with auth, database, and secrets plugin types, with the following logic to populate `RunningVersion`:

* If a plugin has a `Version()` method implemented, then that is used
* If not, and the plugin is built into the Vault binary, then the `go.mod` version is used
* Otherwise, the it will be the empty string.

My apologies for the length of this PR.

tvoran commented:
For the failing cache tests, some of them are using the PassthroughBackend, so I wonder if that needs another type assertion in request_handling.go:

```diff
diff --git a/vault/request_handling.go b/vault/request_handling.go
index b37085c83c..95243bb38a 100644
--- a/vault/request_handling.go
+++ b/vault/request_handling.go
@@ -1076,7 +1076,7 @@ func (c *Core) handleRequest(ctx context.Context, req *logical.Request) (retResp
                                return nil, auth, retErr
                        }
 
-                       if ptbe, ok := matchingBackend.(*PassthroughBackend); ok {
+                       if ptbe, ok := matchingBackend.(builtinVersionBackend).backend.(*PassthroughBackend); ok {
                                if !ptbe.GeneratesLeases() {
                                        registerLease = false
                                        resp.Secret.Renewable = false
```

tomhjp commented on sdk/plugin/pb/backend.proto:
```diff
@@ -453,6 +459,9 @@ service Backend {
 	rpc Type(Empty) returns (TypeReply);
+
+	// Version returns version information for the plugin.
+	rpc Version(Empty) returns (VersionReply);
```
Adding the method here has a couple of (minor) drawbacks:

1. We can't use a single protobuf definition for both db and auth/secret engines.
2. The original intention was that self-reporting the plugin's version would be opt-in. Anyone using this version of the SDK will have to implement the method to fulfil the `Backend` interface. I suppose they could return an empty version though.

I haven't thought too much about the benefits of the way it is now though - what's the motivation to add it here?

tomhjp replied:
SGTM, although the points above are definitely minor if you think there's good opportunity to simplify by adding to the existing interfaces.

tomhjp commented on sdk/plugin/pb/backend.proto:
```diff
@@ -403,6 +403,12 @@ message TypeReply {
 
+// VersionReply is the reply for the Version method.
+message VersionReply {
+	string name = 1;
```
`name` feels a bit too ambiguous to me, is this the plugin name? Maybe worth being a bit more verbose?

```suggestion
	string plugin_name = 1;
```

tomhjp commented on sdk/plugin/grpc_backend_client.go:
```diff
@@ -280,3 +280,14 @@ func (b *backendGRPCPluginClient) Type() logical.BackendType {
+func (b *backendGRPCPluginClient) Version() logical.VersionInfo {
+	reply, err := b.client.Version(b.doneCtx, &pb.Empty{})
+	if err != nil {
+		b.Logger().Debug("Error getting plugin version; assuming method is missing", "err", err)
```
Is there a specific error we can check for to determine that it's because the method was missing?

swenson commented:
Thanks @tvoran for the tip about the passthroughbackend -- that was exactly the problem.

swenson commented on sdk/logical/logical.go:
```diff
@@ -137,3 +141,12 @@ type Auditor interface {
+
+var (
+	EmptyVersion   = VersionInfo{""}
+	BuiltinVersion = VersionInfo{version.GetVersion().Version + "+builtin"}
```
Oh, awesome. I'll switch to that.

swenson commented:
I converted this into a draft since I have a bit more work to do before it's quite ready. I went ahead and removed the plugin wrapper and used the build information to get the plugin version, as Tom suggested, which simplifies things a little.

swenson commented:
@tvoran @tomhjp this is ready for review, I believe. I think I addressed all of your comments, and I did a lot of testing.

swenson commented on vault/plugin_catalog.go:
```diff
@@ -874,42 +869,3 @@ func isPluginType(s string) bool {
 	return err == nil
 }
-
-func (c *PluginCatalog) getBuiltinVersion(pluginType consts.PluginType, pluginName string) string {
```
I moved this function to more central place so it could be used more easily. It didn't really need access to the `pluginCatalog` to work.

tomhjp commented on builtin/logical/database/path_config_connection.go:
```diff
@@ -394,6 +394,11 @@ func (b *databaseBackend) connectionWriteHandler() framework.OperationFunc {
+		// don't override a manually specified plugin version
+		// but if it is empty, then add in the self-reported version for informational purposes
+		if config.PluginVersion == "" {
+			config.PluginVersion = dbw.Version().Version
```
I think this could cause issues the next time the config is updated if it's different to the version we did the initial lookup with, as next time we'll lookup the reported version and maybe get a different entry from the catalog. Could we call this method as we're registering the plugin, and either a) Use the reported version in the catalog if none was provided or b) Error if the reported version is different from the one provided? Then we shouldn't need to make any updates here as the catalog and reported versions will always match up.

It might be interesting to query the version from the plugin and validate that it's the version we think it should be or error out, though I can't think of any reason that would fail if we validate during plugin registration.

swenson replied:
It is definitely possible to add a check between `PluginVersion` and the self-reported version on registration.

I primarily added this because, if no version is specified on registration, it is otherwise impossible to tell what version of the plugin is running, and I wanted to expose this somewhere.

It doesn't even really make sense to have this information in the config at all, but there's no mount or auth table for database configs.

tomhjp commented on builtin/plugin/backend.go:
```diff
@@ -79,22 +79,44 @@ func Backend(ctx context.Context, conf *logical.BackendConfig) (*PluginBackend,
-	b.Backend = &framework.Backend{
-		PathsSpecial: paths,
-		BackendType:  btype,
+	b.Backend = &placeholderBackend{
```
`placeholderBackend.IsExternal()` always returns true, but the function comment for `Backend` implies the backend could be builtin. Not sure if I'm just misreading though?

swenson replied:
Ah, you're right, the placeholder is used even for builtin plugins. I'll see if I can just forward the `IsExternal` method.

tomhjp commented on changelog/17088.txt:
````diff
@@ -0,0 +1,3 @@
+```release-note:feature
+plugins: Adding version to plugin GRPC interface
````
Formatting is slightly different for feature changelogs:

````
```release-note:feature
**Feature Name**: Description of feature - for example "Custom password policies are now supported for all database engines."
```
````

swenson replied:
It's more of an improvement really, so I'll just change that :)

tomhjp commented on sdk/logical/version.proto:
```diff
@@ -0,0 +1,18 @@
+}
+
+// Versioned is an optional RPC service implemented by plugins.
+service Versioned {
```
WDYT about `PluginVersion` to differentiate from Vault version and protocol version?

swenson replied:
:+1:

tomhjp commented on sdk/logical/version.proto:
```diff
@@ -0,0 +1,18 @@
+// VersionReply is the reply for the Version method.
+message VersionReply {
+  string plugin_name = 1;
+  string version = 2;
```
I originally wanted to use version everywhere, so started sprinkling that around. But then we've found a couple of places where we need to use plugin_version to avoid ambiguity. I'm wondering if we should standardise on plugin_version everywhere for consistency, and for an abundance of clarity?

swenson replied:
I think that makes sense. I'll rename it.

tomhjp commented on sdk/plugin/grpc_backend_client.go:
```diff
@@ -280,3 +280,23 @@ func (b *backendGRPCPluginClient) Type() logical.BackendType {
+				return logical.EmptyVersion
+			}
+		}
+		b.Logger().Warn("Unknown error getting plugin version", "err", err)
```
Something would have to go seriously wrong for this to happen, which makes me a bit nervous to only log it server-side and continue. I see `Type` does something similar above, but perhaps it's worth adding an error return value to our new interface signature before we first ship it so that we have an easier time if we have need for it at a later date. WDYT?

swenson replied:
I was trying to mimic `Type`'s relative safety -- worst case scenario, the `Version` will just return `""`. I can think of few reasons why this might fail, like if the plugin dies or is running too slowly.

(`Type` won't even log a warning if an error occurs. :))

I'm open to adding an error return though; I was just trying to keep it as simple as possible.

tomhjp replied:
Yeah it's definitely nice to have that simplicity. I was thinking a little of possible future changes to the plugin architecture, e.g. what if we wanted to support remote plugins where transport errors might become less unusual. It's perfectly possible the error will never be a particularly useful part of the API though, so I'm not 100% about it.

tomhjp commented on sdk/plugin/grpc_backend_server.go:
```diff
@@ -266,3 +267,19 @@ func (b *backendGRPCPluginServer) Type(ctx context.Context, _ *pb.Empty) (*pb.Ty
+			Version: versioner.Version().Version,
+		}, nil
+	}
+	return &logical.VersionReply{
```
Use `logical.EmptyVersion`?

tomhjp replied:
Ah thanks, my mistake 👍

tomhjp commented on vault/auth.go:
```diff
@@ -182,7 +183,16 @@ func (c *Core) enableCredentialInternal(ctx context.Context, entry *MountEntry,
+	}
+	if entry.RunningVersion == "" {
+		// don't set the running version to a builtin if it is running as an external plugin
+		if externaler, ok := backend.(logical.Externaler); !ok || !externaler.IsExternal() {
```
When would `ok` be false, is that more a defensive check to avoid unexpected panics?

swenson replied:
It's mostly defensive. There are so many plugin middlewares we use, and if we end up adding a new one in the future, I wouldn't want this to panic.

tomhjp replied:
Gotcha, seems prudent 👍

tomhjp reviewed (commented):
Looking good! It might be that my comment about validating the self-reported version during registration would be best left for another PR if that's non-trivial

swenson commented:
@tomhjp I think I addressed your concerns. PTAL.

tomhjp reviewed (approved):
LGTM!

tvoran commented on builtin/plugin/backend.go:
```diff
@@ -79,22 +79,52 @@ func Backend(ctx context.Context, conf *logical.BackendConfig) (*PluginBackend,
+	}
+
+	external := false
+	if externaler, ok := raw.(logical.Externaler); ok {
+		external = externaler.IsExternal()
+	}
```
Are builtin plugins ever external?

swenson replied:
No, for our purposes. Here, external means GRPC. A builtin plugin could be built as an external plugin and run as a separate process. This is the exact case where we don't want to query the `go.mod` / debug information for the version, since the external version might be different (e.g., to fix a bug).

tvoran commented on sdk/framework/backend.go:
```diff
@@ -91,6 +91,9 @@ type Backend struct {
 	BackendType logical.BackendType
 
+	// RunningVersion is the optional version that will be self-reported
+	RunningVersion string
```
Is this all that needs to be set in an external plugin once it's using the new sdk version? Like in the [newBackend() function](https://github.com/hashicorp/vault-plugin-secrets-kubernetes/blob/b423467b11688b76d1a873e806566f153ab51fbd/backend.go#L71), for instance?

swenson replied:
Yes. If a plugin sets this value, then that will be sufficient for the running version to appear in the mount/auth table.

They could also override the method directly, e.g., I tested this with vault-plugin-auth-azure:

```go
var _ logical.PluginVersioner = &azureAuthBackend{}

func (_ *azureAuthBackend) PluginVersion() logical.PluginVersion {
	return logical.PluginVersion{
		Version: "v1234",
	}
}
```

tvoran reviewed (commented):
Just a couple naive questions.

swenson commented:
Thanks!
